### ISSUE TITLE
CBORParser seems to promote float to double

### ISSUE THREAD

**Issue by heligo** [FasterXML/jackson-dataformats-binary#32]
I want to use Jackson CBOR because the CBOR RFC has self describing types and distinguishes between different sizes of floating point and integral numbers. I want to use this feature of the format specification to infer a schema from collections of CBOR messages. 

By default, the CBORGenerator will write longs < Integer.MAX_VALUE as ints, but this can be disabled. 

CBORGenerator writes floats and doubles properly. CBORParser correctly identifies floats as floats (nextToken case 26) but constructs a double nevertheless. This seems to be because JsonToken in Jackson core does not differentiate between floats and doubles.

Here's a simple test to demonstrate the issue:
```java @Test
  public void testPrimitiveTypeInvariance() throws IOException {
    ObjectMapper mapper = new ObjectMapper(
            new CBORFactory().disable(CBORGenerator.Feature.WRITE_MINIMAL_INTS)
    );
    Map<String, Object> map = ImmutableMap.of("longField", 1L, "intField", 1, "doubleField", 1.0, "floatField", 1.0f);
    byte[] json = mapper.writeValueAsBytes(map);
    Map<String, Object> fromCbor = mapper.readerFor(Map.class).readValue(json);
    test(fromCbor);
  }

  private void test(Map<String, Object> map) {
    Assert.assertTrue("long not preserved: " + className(map.get("longField")), map.get("longField") instanceof Long);
    Assert.assertTrue("int not preserved: " + className(map.get("intField")), map.get("intField") instanceof Integer);
    Assert.assertTrue("double not preserved: " + className(map.get("doubleField")), map.get("doubleField") instanceof Double);
    Assert.assertTrue("float not preserved: " + className(map.get("floatField")), map.get("floatField") instanceof Float);
  }


  private String className(Object o) {
    if(null == o) return "null";
    return o.getClass().getCanonicalName();
  }
```

The version of jackson-dataformat-cbor is 2.8.3.

**Comment by cowtowncoder** [comment]
This sounds familiar. Issue #28 was for Avro backend, and for `NumberNode`.

Thank you for reporting the problem. It should be solvable since although same token is used for both, underlying physical type can be determined (via `getNumberType()`).

**Comment by cowtowncoder** [comment]
Looks like this is a problem with `jackson-databind`; need to create an issue there, to be fixed in 2.8.6 (2.8.5 was just released).

**Comment by heligo** [comment]
Nice work. Would it be helpful if I verify somehow?

**Comment by cowtowncoder** [comment]
Sure. Fix is in for `jackson-databind`, branch 2.8 (and `master` too), so as long as you compile locally (or use Snapshot via sonatype snapshot repo) you should be able to verify.

**Comment by cowtowncoder** [comment]
Added a test in `master`, passes.
